vm/state_migration/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+    MigratorNotFound(Cid),
+    #[error("Failed updating new actor state: {0}")]
+    SetActorState(String),
+    #[error("State tree creation failed")]
+    StateTreeCreation(String),
+    #[error("Incomplete migration specification with {0} code CIDs")]
+    IncompleteMigrationSpec(usize),
+    #[error("Thread pool creation failed: {0}")]
+    ThreadPoolCreation(ThreadPoolBuildError),
+    #[error("Migration failed")]
+    Other,
+}
+
+pub struct StateMigration<BS> {
+    migrations: HashMap<Cid, Migrator<BS>>,
+    deferred_code_ids: HashSet<Cid>,
+}
+
+impl<BS: BlockStore + Send + Sync> Default for StateMigration<BS> {
+    fn default() -> Self {
+        let mut migrations = HashMap::new();
+        migrations.insert(
+            *actorv3::ACCOUNT_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::ACCOUNT_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::CRON_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::CRON_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::INIT_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::INIT_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::MULTISIG_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::MULTISIG_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::PAYCH_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::PAYCH_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::REWARD_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::REWARD_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::MARKET_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::MARKET_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::POWER_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::POWER_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::SYSTEM_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::SYSTEM_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::VERIFREG_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::VERIFREG_ACTOR_CODE_ID),
+        );
+
+        Self {
+            migrations,
+            deferred_code_ids: HashSet::new(),
+        }
+    }
+}
+
+impl<BS: BlockStore + Send + Sync> StateMigration<BS> {
+    pub fn add_migrator(&mut self, prior_cid: Cid, migrator: Migrator<BS>) {
+        self.migrations.insert(prior_cid, migrator);
+    }
+
+    pub fn migrate_state_tree(
+        &mut self,
+        store: Arc<BS>,
+        prior_epoch: ChainEpoch,
+        actors_in: StateTree<BS>,
+        mut actors_out: StateTree<BS>,
+    ) -> MigrationResult<Cid> {
+        if self.migrations.len() + self.deferred_code_ids.len() != ACTORS_COUNT {
+            return Err(MigrationError::IncompleteMigrationSpec(
+                self.migrations.len(),
+            ));
+        }
+
+        let cpus = num_cpus::get();
+        let chan_size = 2;
```

Comment:
really? this was the optimal size? i feel like this should also be a fucntion of num cpus, no?

Reply:
Ah no this should be 4 (will do the change). Anything greater than that and i see performance degrade. With 4 we're seeing 13-15 secs.

Reply:
Then would you say the optimal is num_cpus/2?

Reply:
mm, right. Let's do that

Reply:
If the ideal is 4, why not do `std::cmp::min(4, num_cpus::get() - 1);` ? Subtracting one core so the system doesn't lock up.